### ISSUE TITLE
How to handle "Could not load snapshot text for XXX" warning?

### ISSUE THREAD

**Issue by tribbloid** [scalameta/metals#7064]
### Describe the bug

When trying to compile a Scala 2 project:

https://github.com/tribbloid/prover-commons

I got a lot of errors like this:

![image](https://github.com/user-attachments/assets/8a930926-3b90-4196-83c2-e859c837a07c)

```
2024.12.23 22:16:39 WARN  Could not load snapshot text for /home/peng/git/spookystuff/prover-commons/module/core/src/main/scala/ai/acyclic/prover/commons/graph/local/ops/AnyGraphUnary.scala
```

What does it mean? How to fix it?

### Expected behavior

Should load snapshot text, or at least direct user on how to find it

### Operating system

Linux

### Editor/Extension

VS Code

### Version of Metals

1.4.2

**Comment by tribbloid** [comment]
this problem is also encountered in IntelliJ IDEA + LSP4IJ plugin:

```

2024.12.27 17:58:36 WARN  Could not load snapshot text for /home/peng/git-template/scaffold-gradle-kts/core/src/main/scala/ai/acyclic/scaffold/Example.scala
2024.12.27 17:58:36 WARN  Could not load snapshot text for /home/peng/git-template/scaffold-gradle-kts/core/src/main/scala/ai/acyclic/scaffold/Example.scala
2024.12.27 17:58:36 WARN  Could not load snapshot text for /home/peng/git-template/scaffold-gradle-kts/core/src/main/scala/ai/acyclic/scaffold/Example.scala
```

could be easily reproduced on a sample project: https://github.com/tribbloid/scaffold-gradle-kts/tree/b421ae8cb778affaf3759a9b79b9462e161dd9ed

**Comment by tribbloid** [comment]
LSP appears to have invoked gradle-bloop 1.6.2 before launching, could it be where problem come from?

**Comment by tgodzik** [comment]
That should not be the problem, this only happens if the currently compiled semanticdb is not generated from the current file.

My guess is that this might be caused by checking out a different branch and code not compiling anymore. In that case if the file was never opened we never did save the text used for semanticdb.

I need to make it more foolproof, since this logic seem quite flaky. This should go away once the code is fully compiled, but we do have some fallbacks so this should no longer be that big of an issue.

**Comment by tribbloid** [comment]
ah that explained it, almost forgot that Scala 3 best effort compilation feature is not available for Scala 2

in this case it should be swept into the backlog

**Comment by tgodzik** [comment]
We should only print this warning if none of the fallbacks worked, or better yet create a report for the user to bring to us. The current warning is less than useful

**Comment by matthughes** [comment]
I get this all the time FWIW.  Scala 3.7.4.

**Comment by tgodzik** [comment]
Curious, is it for a file that does exist and compiles? It should not be a big issue currently, but it would be good to figure out why it's happening. Or maybe you have a repo that this happens in?

**Comment by matthughes** [comment]
Yes it's for a file that exists and compiles.  It's so weird.  I had a line
like:

val x: Instant = ???

First time I tried to do code action on Instant to get the import, it
worked and provided the import.  If I don't accept and then try again, you
get this error.  I've definitely seen this many times on other files.  This
and

```empty definition using pc, found symbol in pc: `<error>`#`<error>`.
empty definition using semanticdb empty definition using fallback non-local
guesses:```

Make up most of my error reports.

**Comment by tgodzik** [comment]
my current guess is that somehow the md5 calculated for semanticdb is different than the one we calculate locally, but I can't reproduce it reliably. Metals v2 should have less problems with it, but it would be great to figure out why the discrepancy. It's probably something with newlines or other whitespace.
